File: packages/scelo-r/R/workspace.R
# Workspace: the Hard Data layer's "global workspace" diagnostics.
#
# A port of the IDE's numpy bottleneck bridge (apps/web/.../bridges/
# bottleneckPython.ts) and the active-subspace pieces around it: which few
# directions in the drivers the report heads actually turn on, how sparse
# and non-negative the broadcast from codes to heads is, and whether the
# codes are causally aligned with the marginal slopes. The linear special
# case with one code is exactly Lee-Carter; see Denewade (2026), "A Global
# Workspace for Actuarial Models". The eigen-decompositions are base R's
# eigen(); the projected-gradient fit of the broadcast runs on all heads at
# once (one matrix update per step instead of one per head), which gives
# the numpy bridge's numbers in a fraction of its time.

#' Participation ratio
#'
#' (sum of lambda)^2 / sum of lambda^2: the effective number of directions.
#' @param eigenvalues A numeric vector (non-positive entries are ignored).
#' @return A number (0 for no positive eigenvalue).
#' @examples
#' sc_participation_ratio(c(3, 1, 0.1))
#' @export
sc_participation_ratio <- function(eigenvalues) {
  w <- as.numeric(eigenvalues)
  w <- w[!is.na(w) & w > 0]
  if (!length(w)) return(0)
  sum(w)^2 / sum(w^2)
}

# Name a direction by its largest loadings: "mortality trend up, cohort effect down".
.sc_name_code <- function(loadings, cols) {
  ord <- order(-abs(loadings))
  top <- abs(loadings[ord[1]])
  parts <- character()
  for (j in utils::head(ord, 3)) {
    if (top > 0 && abs(loadings[j]) >= 0.35 * top) {
      parts <- c(parts, paste(gsub("[_-]", " ", as.character(cols[j])), if (loadings[j] > 0) "up" else "down"))
    }
  }
  if (length(parts)) paste(parts, collapse = ", ") else "mixed"
}

# Numeric matrix of the named columns, complete rows only, thinned to max_rows
# by a regular stride (the IDE's cap).
.sc_numeric_matrix <- function(df, cols, max_rows) {
  X <- as.data.frame(lapply(df[cols], function(v) suppressWarnings(as.numeric(if (is.factor(v)) as.character(v) else v))))
  X <- X[stats::complete.cases(X), , drop = FALSE]
  if (nrow(X) > max_rows) {
    step <- max(1L, nrow(X) %/% max_rows)
    X <- X[seq(1L, nrow(X), by = step), , drop = FALSE]
  }
  m <- as.matrix(X)
  dimnames(m) <- NULL
  m
}

.sc_standardise <- function(X) {
  mu <- colMeans(X)
  sd <- apply(X, 2, stats::sd)
  sd[sd < 1e-9] <- 1
  sweep(sweep(X, 2, mu, "-"), 2, sd, "/")
}

#' Workspace bottleneck
#'
#' `r` codes (top eigen-directions of the standardised drivers) broadcast
#' non-negatively to every column. Returns the heads x codes broadcast
#' matrix with the code names, and as attributes the participation ratio,
#' reconstruction R^2, causal alignment and sparsity (the IDE's four
#' workspace metrics).
#' @param df A data frame.
#' @param columns Driver columns (all numeric columns by default).
#' @param r Number of codes.
#' @param l1 L1 penalty of the projected-gradient fit.
#' @param max_rows Thin the frame to at most this many rows.
#' @return A `scelo_table` with a `head` column and one `code k: name` column
#'   per code; attributes `participation_ratio`, `reconstruction_r2`,
#'   `causal_alignment`, `sparsity`, `code_loadings` (a data frame of the
#'   oriented eigenvectors), `eigenvalues`.
#' @examples
#' ws <- sc_sample("workspace-demo")
#' b <- sc_bottleneck(ws[, 1:6], r = 2)
#' attr(b, "causal_alignment")
#' @export
sc_bottleneck <- function(df, columns = NULL, r = 3, l1 = 1e-3, max_rows = 20000) {
  .sc_tool("sc_bottleneck", list(df = df, columns = columns, r = r, l1 = l1), df, {
    cols <- if (!is.null(columns)) as.character(columns) else .sc_numeric_columns(df)
    X <- .sc_numeric_matrix(df, cols, max_rows)
    if (nrow(X) < 10 || ncol(X) < 3) stop("need at least 10 complete rows and 3 numeric columns", call. = FALSE)
    p <- ncol(X)
    r <- max(1L, min(as.integer(r), p - 1L))
    Z <- .sc_standardise(X)
    C <- stats::cov(Z)
    e <- eigen(C, symmetric = TRUE)
    w <- e$values
    Vr <- e$vectors[, seq_len(r), drop = FALSE]
    codes <- Z %*% Vr
    rowsum <- rowSums(Z)
    for (k in seq_len(r)) {
      if (stats::sd(rowsum) > 0 && stats::sd(codes[, k]) > 0 && stats::cor(codes[, k], rowsum) < 0) {
        Vr[, k] <- -Vr[, k]
        codes[, k] <- -codes[, k]
      }
    }
    G <- crossprod(codes)
    lr <- 1 / (sum(diag(G)) + 1e-9)
    Gt <- crossprod(Z, codes) # p x r: codes' z_c for every head at once
    B <- matrix(0, p, r)
    for (i in seq_len(300)) B <- pmax(B - lr * (B %*% G - Gt + l1), 0)
    recon <- codes %*% t(B)
    ss_res <- colSums((Z - recon)^2)
    ss_tot <- colSums(Z^2)
    r2 <- mean(pmin(1, pmax(0, 1 - ss_res / ifelse(ss_tot > 0, ss_tot, 1))))
    aligns <- numeric()
    for (k in seq_len(r)) {
      zk <- codes[, k]
      vk <- mean((zk - mean(zk))^2)
      slopes <- if (vk > 0) as.numeric(stats::cov(Z, zk)) / vk else rep(0, p)
      if (stats::sd(B[, k]) > 0 && stats::sd(slopes) > 0) aligns <- c(aligns, stats::cor(B[, k], slopes)^2)
    }
    align <- if (length(aligns)) mean(aligns) else 0
    pr <- sc_participation_ratio(w[seq_len(r)])
    sparsity <- if (length(B) && max(abs(B)) > 0) mean(abs(B) < 0.02 * max(abs(B))) else 1
    names <- vapply(seq_len(r), function(k) .sc_name_code(Vr[, k], cols), character(1))
    out <- data.frame(head = cols, stringsAsFactors = FALSE)
    for (k in seq_len(r)) out[[sprintf("code %d: %s", k, names[k])]] <- B[, k]
    t <- sc_table(out, title = sprintf("Workspace bottleneck · %d codes · %s rows", r, format(nrow(X), big.mark = ",")),
                  basis = sprintf("PR %.2f · reconstruction R² %.2f · causal alignment %.2f · sparsity %.2f", pr, r2, align, sparsity), stage = "hard", notes = c(
      "Codes are the leading eigenvectors of the standardised driver covariance, oriented positively; the broadcast B ≥ 0 is fitted by projected gradient with an L1 penalty (300 steps).",
      "Participation ratio = effective number of codes; causal alignment = how well each code's broadcast matches the marginal slopes of the heads on that code."
    ))
    loadings <- as.data.frame(Vr)
    names(loadings) <- names
    loadings <- cbind(data.frame(head = cols, stringsAsFactors = FALSE), loadings)
    attr(t, "participation_ratio") <- pr
    attr(t, "reconstruction_r2") <- r2
    attr(t, "causal_alignment") <- align
    attr(t, "sparsity") <- sparsity
    attr(t, "code_loadings") <- loadings
    attr(t, "eigenvalues") <- w
    t
  })
}

#' Active subspace of a readout
#'
#' Eigen-directions of the gradient covariance C = E(grad f grad f') of a
#' linear-quadratic surrogate of the readout. Reports each direction's
#' sensitivity share, input-variance share and named loadings: a direction
#' can carry most of the decision and almost none of the variance (the
#' workspace signature).
#' @param df A data frame.
#' @param readout The output column.
#' @param drivers Driver columns (every other numeric column by default).
#' @param max_rows Thin the frame to at most this many rows.
#' @return A `scelo_table` with `direction`, `eigenvalue`, `sensitivity_share`,
#'   `variance_share`, `name`, `loadings` (up to six directions); attributes
#'   `rank`, `participation_ratio`, `surrogate_r2`, `sensitivity_spectrum`,
#'   `variance_spectrum`, `directions` (the eigenvector matrix).
#' @examples
#' ws <- sc_sample("workspace-demo")
#' a <- sc_active_subspace(ws, "annuity_60", drivers = c("mortality_trend", "cohort_effect", "smoking_index", "crude_rate"))
#' attr(a, "rank")
#' @export
sc_active_subspace <- function(df, readout, drivers = NULL, max_rows = 20000) {
  .sc_tool("sc_active_subspace", list(df = df, readout = readout, drivers = drivers), df, {
    if (!readout %in% names(df)) stop(sprintf('column "%s" is not in the data (have: %s)', readout, paste(names(df), collapse = ", ")), call. = FALSE)
    cols <- if (!is.null(drivers)) as.character(drivers) else setdiff(.sc_numeric_columns(df), readout)
    M <- .sc_numeric_matrix(df, c(cols, readout), max_rows)
    X <- M[, seq_along(cols), drop = FALSE]
    y <- M[, length(cols) + 1L]
    Z <- .sc_standardise(X)
    n <- nrow(Z)
    p <- ncol(Z)
    # quadratic surrogate: y ~ b0 + sum b_i z_i + sum_{i<=j} c_ij z_i z_j
    pairs <- which(upper.tri(diag(p), diag = TRUE), arr.ind = TRUE)
    pairs <- pairs[order(pairs[, 1], pairs[, 2]), , drop = FALSE]
    Fm <- cbind(1, Z, Z[, pairs[, 1], drop = FALSE] * Z[, pairs[, 2], drop = FALSE])
    fit <- stats::lm.fit(Fm, y)
    beta <- fit$coefficients
    beta[is.na(beta)] <- 0
    yhat <- as.numeric(Fm %*% beta)
    r2 <- if (stats::var(y) > 0) 1 - sum((y - yhat)^2) / sum((y - mean(y))^2) else 0
    lin <- beta[1 + seq_len(p)]
    Q <- matrix(0, p, p) # symmetric: 2 c_ii on the diagonal, c_ij off it
    for (k in seq_len(nrow(pairs))) {
      i <- pairs[k, 1]; j <- pairs[k, 2]
      if (i == j) Q[i, i] <- 2 * beta[1 + p + k] else { Q[i, j] <- beta[1 + p + k]; Q[j, i] <- beta[1 + p + k] }
    }
    grads <- matrix(lin, n, p, byrow = TRUE) + Z %*% Q
    Cf <- crossprod(grads) / n
    e <- eigen(Cf, symmetric = TRUE)
    w <- e$values
    V <- e$vectors
    CZ <- stats::cov(Z)
    var_eig <- sort(eigen(CZ, symmetric = TRUE, only.values = TRUE)$values, decreasing = TRUE)
    tau <- if (w[1] > 0) 1e-3 * w[1] else 0
    rank <- sum(w > tau)
    kk <- seq_len(min(6L, p))
    rows <- lapply(kk, function(k) {
      v <- V[, k]
      vshare <- if (sum(var_eig) > 0) as.numeric(t(v) %*% CZ %*% v) / sum(var_eig) else 0
      keep <- abs(v) >= 0.2
      data.frame(direction = k, eigenvalue = w[k], sensitivity_share = if (sum(w) > 0) w[k] / sum(w) else 0, variance_share = vshare,
                 name = .sc_name_code(v, cols), loadings = paste(sprintf("%s:%+.2f", cols[keep], v[keep]), collapse = ", "), stringsAsFactors = FALSE)
    })
    out <- do.call(rbind, rows)
    pr <- sc_participation_ratio(w)
    t <- sc_table(out, title = sprintf("Active subspace · %s", readout), basis = sprintf("%d drivers · surrogate R² %.2f · rank %d · PR %.2f", p, r2, rank, pr), stage = "hard", notes = c(
      "Directions are eigenvectors of C = E[∇f∇fᵀ] for a quadratic surrogate of the readout; sensitivity share is the share of C's trace, variance share the share of input variance the direction occupies.",
      sprintf("Workspace variance fraction %.3f over the %d active directions.", sum(utils::head(out$variance_share, rank)), rank)
    ))
    attr(t, "rank") <- rank
    attr(t, "participation_ratio") <- pr
    attr(t, "surrogate_r2") <- r2
    attr(t, "sensitivity_spectrum") <- w
    attr(t, "variance_spectrum") <- var_eig
    attr(t, "directions") <- V
    t
  })
}

File: packages/scelo-py/src/scelo/workspace.py
"""Workspace: the Hard Data layer's "global workspace" diagnostics.

A port of the IDE's numpy bottleneck bridge (apps/web/.../bridges/
bottleneckPython.ts) and the active-subspace pieces around it: which few
directions in the drivers the report heads actually turn on, how sparse and
non-negative the broadcast from codes to heads is, and whether the codes are
causally aligned with the marginal slopes. The linear special case with one
code is exactly Lee–Carter; see Denewade (2026), "A Global Workspace for
Actuarial Models".
"""

from __future__ import annotations

import math
from typing import Any, Dict, List, Optional, Sequence

import numpy as np
import pandas as pd

from ._alias import numeric_columns
from ._audit import tool
from ._table import Table

__all__ = ["bottleneck", "active_subspace", "participation_ratio"]


def participation_ratio(eigenvalues: Sequence[float]) -> float:
    """(Σλ)² / Σλ²: the effective number of directions."""
    w = np.asarray(eigenvalues, dtype=float)
    w = w[w > 0]
    return float(w.sum() ** 2 / np.sum(w ** 2)) if w.size else 0.0


def _name_code(loadings: np.ndarray, cols: Sequence[str]) -> str:
    order = np.argsort(-np.abs(loadings))
    top = abs(loadings[order[0]])
    parts = []
    for j in order[:3]:
        if abs(loadings[j]) >= 0.35 * top and top > 0:
            parts.append(f"{str(cols[j]).replace('_', ' ').replace('-', ' ')} {'up' if loadings[j] > 0 else 'down'}")
    return ", ".join(parts) if parts else "mixed"


@tool
def bottleneck(df: pd.DataFrame, columns: Optional[Sequence[str]] = None, r: int = 3, l1: float = 1e-3, max_rows: int = 20_000) -> Table:
    """Workspace bottleneck: r codes (top eigen-directions of the standardised drivers) broadcast non-negatively to every column.

    Returns the heads × codes broadcast matrix with the code names, and in
    ``attrs``: participation ratio, reconstruction R², causal alignment and
    sparsity (the IDE's four workspace metrics).
    """
    cols = list(columns) if columns is not None else numeric_columns(df)
    X = df[cols].apply(pd.to_numeric, errors="coerce").dropna()
    if len(X) > max_rows:
        X = X.iloc[:: max(1, len(X) // max_rows)]
    X = X.to_numpy(dtype=float)
    if X.shape[0] < 10 or X.shape[1] < 3:
        raise ValueError("need at least 10 complete rows and 3 numeric columns")
    r = max(1, min(r, X.shape[1] - 1))
    mu = X.mean(0)
    sd = X.std(0, ddof=1)
    sd[sd < 1e-9] = 1
    Z = (X - mu) / sd
    C = np.cov(Z, rowvar=False)
    w, V = np.linalg.eigh(C)
    order = np.argsort(w)[::-1]
    w, V = w[order], V[:, order]
    Vr = V[:, :r].copy()
    codes = Z @ Vr
    rowsum = Z.sum(1)
    for k in range(r):
        if rowsum.std() > 0 and codes[:, k].std() > 0 and np.corrcoef(codes[:, k], rowsum)[0, 1] < 0:
            Vr[:, k] *= -1
            codes[:, k] *= -1
    G = codes.T @ codes
    lr = 1 / (np.trace(G) + 1e-9)
    B = np.zeros((X.shape[1], r))
    for c in range(X.shape[1]):
        b = np.zeros(r)
        g = codes.T @ Z[:, c]
        for _ in range(300):
            b = np.maximum(0, b - lr * (G @ b - g + l1))
        B[c] = b
    recon = codes @ B.T
    ss_res = ((Z - recon) ** 2).sum(0)
    ss_tot = (Z ** 2).sum(0)
    r2 = float(np.mean(np.clip(1 - ss_res / np.where(ss_tot > 0, ss_tot, 1), 0, 1)))
    aligns = []
    for k in range(r):
        zk = codes[:, k]
        slopes = np.array([np.cov(Z[:, c], zk)[0, 1] / zk.var() if zk.var() > 0 else 0 for c in range(X.shape[1])])
        if B[:, k].std() > 0 and slopes.std() > 0:
            aligns.append(np.corrcoef(B[:, k], slopes)[0, 1] ** 2)
    align = float(np.mean(aligns)) if aligns else 0.0
    pr = participation_ratio(w[:r])
    sparsity = float(np.mean(np.abs(B) < 0.02 * np.abs(B).max())) if B.size and np.abs(B).max() > 0 else 1.0
    names = [_name_code(Vr[:, k], cols) for k in range(r)]
    out = pd.DataFrame(B, index=pd.Index(cols, name="head"), columns=[f"code {k + 1}: {n}" for k, n in enumerate(names)])
    t = Table(out, title=f"Workspace bottleneck · {r} codes · {X.shape[0]:,} rows", basis=f"PR {pr:.2f} · reconstruction R² {r2:.2f} · causal alignment {align:.2f} · sparsity {sparsity:.2f}", stage="hard", notes=[
        "Codes are the leading eigenvectors of the standardised driver covariance, oriented positively; the broadcast B ≥ 0 is fitted by projected gradient with an L1 penalty (300 steps).",
        "Participation ratio = effective number of codes; causal alignment = how well each code's broadcast matches the marginal slopes of the heads on that code.",
    ])
    t.attrs.update(participation_ratio=pr, reconstruction_r2=r2, causal_alignment=align, sparsity=sparsity, code_loadings=pd.DataFrame(Vr, index=cols, columns=names), eigenvalues=w)
    return t


@tool
def active_subspace(df: pd.DataFrame, readout: str, drivers: Optional[Sequence[str]] = None, *, max_rows: int = 20_000) -> Table:
    """Active subspace of a readout: eigen-directions of the gradient covariance C = E[∇f ∇fᵀ] of a linear-quadratic surrogate.

    Reports each direction's sensitivity share, input-variance share and
    named loadings: a direction can carry most of the decision and almost
    none of the variance (the workspace signature).
    """
    cols = list(drivers) if drivers is not None else [c for c in numeric_columns(df) if c != readout]
    d = df[cols + [readout]].apply(pd.to_numeric, errors="coerce").dropna()
    if len(d) > max_rows:
        d = d.iloc[:: max(1, len(d) // max_rows)]
    X = d[cols].to_numpy(dtype=float)
    y = d[readout].to_numpy(dtype=float)
    mu, sd = X.mean(0), X.std(0, ddof=1)
    sd[sd < 1e-9] = 1
    Z = (X - mu) / sd
    n, p = Z.shape
    # quadratic surrogate: y ~ b0 + Σ b_i z_i + Σ_{i<=j} c_ij z_i z_j
    feats = [np.ones(n)] + [Z[:, i] for i in range(p)] + [Z[:, i] * Z[:, j] for i in range(p) for j in range(i, p)]
    F = np.column_stack(feats)
    beta, *_ = np.linalg.lstsq(F, y, rcond=None)
    yhat = F @ beta
    r2 = float(1 - np.sum((y - yhat) ** 2) / np.sum((y - y.mean()) ** 2)) if y.var() > 0 else 0.0
    # gradient at every row
    lin = beta[1:1 + p]
    quad = np.zeros((p, p))
    idx = 1 + p
    for i in range(p):
        for j in range(i, p):
            quad[i, j] = beta[idx]
            idx += 1
    grads = np.tile(lin, (n, 1))
    for i in range(p):
        for j in range(i, p):
            if i == j:
                grads[:, i] += 2 * quad[i, j] * Z[:, i]
            else:
                grads[:, i] += quad[i, j] * Z[:, j]
                grads[:, j] += quad[i, j] * Z[:, i]
    Cf = grads.T @ grads / n
    w, V = np.linalg.eigh(Cf)
    order = np.argsort(w)[::-1]
    w, V = w[order], V[:, order]
    var_eig = np.sort(np.linalg.eigvalsh(np.cov(Z, rowvar=False)))[::-1]
    tau = 1e-3 * w[0] if w[0] > 0 else 0
    rank = int((w > tau).sum())
    rows = []
    for k in range(min(6, p)):
        vshare = float(V[:, k] @ np.cov(Z, rowvar=False) @ V[:, k] / var_eig.sum()) if var_eig.sum() > 0 else 0.0
        rows.append({"direction": k + 1, "eigenvalue": w[k], "sensitivity_share": w[k] / w.sum() if w.sum() > 0 else 0, "variance_share": vshare, "name": _name_code(V[:, k], cols),
                     "loadings": ", ".join(f"{c}:{V[j, k]:+.2f}" for j, c in enumerate(cols) if abs(V[j, k]) >= 0.2)})
    out = pd.DataFrame(rows)
    pr = participation_ratio(w)
    t = Table(out, title=f"Active subspace · {readout}", basis=f"{p} drivers · surrogate R² {r2:.2f} · rank {rank} · PR {pr:.2f}", stage="hard", notes=[
        "Directions are eigenvectors of C = E[∇f∇fᵀ] for a quadratic surrogate of the readout; sensitivity share is the share of C's trace, variance share the share of input variance the direction occupies.",
        f"Workspace variance fraction {sum(r_['variance_share'] for r_ in rows[:rank]):.3f} over the {rank} active directions.",
    ])
    t.attrs.update(rank=rank, participation_ratio=pr, surrogate_r2=r2, sensitivity_spectrum=w, variance_spectrum=var_eig, directions=V)
    return t
